src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs

```diff
@@ -1242,6 +1242,16 @@ public virtual MethodSymbol TupleUnderlyingMethod
             }
         }
 
+        /// <summary>
+        /// Returns a suitable location for this MethodSymbol for a diagnostic to be placed at. This
+        /// is generally just the first entry in <see cref="Symbol.Locations"/>, but it may be a 
+        /// smaller location if the full location is large.  For example, Locations[0] on  lambda 
+        /// symbols covers the entire syntax, which is inconvenient but remains for compatibility.
+        /// For better diagnostics quality, <see cref="DiagnosticLocation"/> will point to the 
+        /// "delegate" or the "=>".
+        /// </summary>
+        internal virtual Location DiagnosticLocation => Locations[0];
```

Comment:
I'm not certain :)   How does your code today know that it's safe to call Locations[0]  on a methodsymbol?

--

Ok, i just looked throuhg all the the subclasses of MethodSymbol to see how Locations works for them.  It looks like there are two cases where calling DiagnosticLocation would not work (though calling Locations[0] would also not work).  Specifically, "ErrorMethodSymbol" (which has an empty list) and SignatureOnlyMethodSymbol (which throws if you even call Locatoins on it).  

The latter is used only for signature comparison, and it looks like that code knows to not use any sort of location information.  And the former is used to represent bad-calls, and it seems like it must be the case that the compiler either never uses that in codepaths that then later report errors.  Or the compiler codepaths are hardened to not assume non-empty locations for method symbols.

--

Let me turn this in another direction: what would be your preference on how we best approach this problem?   A virtual method felt fine to me.  But if there is a concern that it may be misused, i could certainly approach the problem by having the DataFlow pass cast the MethodSyymbol to a LambdaSymbol when appropriate and get teh DiagnosticLocation from that. 

Would that be preferable?  Do you have an alternative you would like? :)